### PR TITLE
Google Cloud Storage Data Source Support.

### PR DESCRIPTION
**Pull request by hidekoji**
# Description

Google Cloud Storage Data Source Support.

# Checklist
Make sure you have performed the following items before submitting this pull request.
If not, please describe the reason.  

- [ ] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -279,6 +285,7 @@ export(getGoogleBigQueryDataSets)
 export(getGoogleBigQueryProjects)
 export(getGoogleBigQueryTable)
 export(getGoogleBigQueryTables)
+export(getGoogleTokenForCloudStorage)
```
Comment (hideaki) [inline NAMESPACE:288]: Move it to the right alphabetical location.

File: R/google_cloud_storage.R [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  }
+}
+
+#' API to list items inside a Google Cloud Storage Bucket.
+#' @param bucket
+#' @param prefix
+#' @param delimiter
+#' @export
+listItemsInGoogleCloudStorageBucket <- function(bucket, prefix, delimiter){
+  googleCloudStorageR::gcs_list_objects(bucket = bucket, detail="more", prefix = prefix, delimiter = delimiter)
```
Comment (hideaki) [inline R/google_cloud_storage.R:54]: Be consistent on spacing.

File: R/google_cloud_storage.R [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+      # Looking for error that looks like "http_404 Unspecified error".
+      # This seems to be returned when the file does not exist.
+      stop(paste0('EXP-DATASRC-19 :: ', jsonlite::toJSON(c(bucket)), ' :: There is no file in the Google Cloud Storage bucket that matches with the file name.'))
+    }
+    else {
+      stop(e)
+    }
+  })
+  if (nrow(files) == 0) {
+    stop(paste0('EXP-DATASRC-19 :: ', jsonlite::toJSON(bucket), ' :: There is no file in the Google Cloud Storage bucket that matches with the file name.')) # TODO: escape bucket name.
```
Comment (hideaki) [inline R/google_cloud_storage.R:192]: Check if the result of toJSON is an array expression as expected.

File: R/google_cloud_storage.R [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+  }
+  getCSVFilesFromGoogleCloudStorage(files = files$name, bucket = bucket, for_preview = for_preview, delim = delim, quote = quote,
+                    col_names = col_names, col_types = col_types, locale = locale, na = na, quoted_na = quoted_na, comment = comment, trim_ws = trim_ws,
+                    skip = skip, n_max = n_max, guess_max = guess_max, progress = progress)
+
+}
+
+
+#'API that imports a Excel file from AWS S3.
```
Comment (hideaki) [inline R/google_cloud_storage.R:201]: Correct the comment.

File: R/oauth.R [hunk 5]
```diff
@@ -270,5 +270,41 @@ refreshGoogleTokenForBigQuery <- function(tokenFileId){
...
+    )
+  } else {
+    stop("OAuth token is not set for Google Cloud Storage")
+  }
+}
+
+#' API to refresh token
+#' @export
+refreshGoogleTokenForCloudStorage <- function(){
+  getGoogleTokenForBigQuery(FALSE)
```
Comment (hideaki) [inline R/oauth.R:306]: Should be cloud storage.

File: R/system.R [hunk 6]
```diff
@@ -2111,10 +2111,15 @@ downloadDataFromGoogleCloudStorage <- function(bucket, folder, download_dir, tok
 
 #' API to get a list of buckets from Google Cloud Storage
 #' @export
-listGoogleCloudStorageBuckets <- function(project, tokenFileId){
+listGoogleCloudStorageBuckets <- function(project, tokenFileId = "", shouldUseCloudStorageToken = FALSE){
```
Comment (hideaki) [inline R/system.R:2114]: change argument name to "service".

File: R/system.R [hunk 7]
```diff
@@ -2220,7 +2225,7 @@ executeGoogleBigQuery <- function(project, query, destinationTable, pageSize = 1
 
 #' API to get projects for current oauth token
 #' @export
-getGoogleBigQueryProjects <- function(tokenFileId=""){
+getGoogleBigQueryProjects <- function(tokenFileId="", shouldUseCloudStorageToken = FALSE){
```
Comment (hideaki) [inline R/system.R:2228]: change argument name to "service".

File: R/google_cloud_storage.R (hunk 3 above)
Comment (hidekoji) [reply R/google_cloud_storage.R:192]: Yes, it's an array expression.

```r
> jsonlite::toJSON("bucket1")
["bucket1"] 
```

File: R/oauth.R (hunk 5 above)
Comment (hidekoji) [reply R/oauth.R]: Fixed it.

File: R/google_cloud_storage.R (hunk 4 above)
Comment (hidekoji) [reply R/google_cloud_storage.R]: Fixed the comment.

File: R/google_cloud_storage.R (hunk 2 above)
Comment (hidekoji) [reply R/google_cloud_storage.R]: fixed the spacing.

File: NAMESPACE (hunk 1 above)
Comment (hidekoji) [reply NAMESPACE]: fixed the order.

**Pull request merged**